rclcpp/include/rclcpp/parameter_client.hpp

```diff
@@ -73,28 +75,33 @@ class AsyncParametersClient
     std::function<void(
       std::shared_future<std::vector<rclcpp::parameter::ParameterVariant>>)> callback = nullptr)
   {
-    std::promise<std::vector<rclcpp::parameter::ParameterVariant>> promise_result;
-    auto future_result = promise_result.get_future().share();
+    auto promise_result =
+      std::make_shared<std::promise<std::vector<rclcpp::parameter::ParameterVariant>>>();
+    this->get_parameters_promises_.insert(promise_result);
+    auto future_result = promise_result->get_future().share();
 
     auto request = std::make_shared<rcl_interfaces::srv::GetParameters::Request>();
     request->names = names;
 
+    auto self = shared_from_this();
```

Comment:
This will only work if the instance is already owned by another `shared_ptr` right? Do we make that guarantee?